### PR TITLE
Removing AlicaEngine::this from TeamObserver

### PR DESCRIPTION
**Pull request by triccyx**
Removing AlicaEngine::this from TeamObserver and subclasses
Still some errors in tests. Some tests need to be changed.

### CODE REVIEW

**Comment by triccyx** [comment]
At the moment I am blocked at:
`test_task_assignment.cpp`
the test load
`MasterPlanTaskAssignment.pml`
that has 4 entrypoints
for this reason 
`RuleBook::initialisationRule`
raises abort `"RB: Masterplan does not have exactly one task!"`

Ant suggestion about this problem? @abhi1287 @veerajsk

**Comment by abhi1287** [comment]
Seems like a mistake. Masterplan can't have more than 1 entry point.  cc: @bjoernschroeder

**Comment by abhi1287** [comment]
I wonder how was it working so far?

**Comment by triccyx** [comment]
Also test `TaskAssignmentTest.switchEntryPoints`
has problems. It uses all the `entryPoints` in `MasterPlanTaskAssignment.pml`. But if I understand well 
the masterplan should contain just one entrypoint

**Comment by bjoernschroeder** [comment]
In rr_devel the tests are working because the engine is not running, and therefore RuleBook::initialisationRule is never called. For these tests, AlicaEngine::start() should not be called, as this will start the planbase, which will later call the initialisation rule in PlanBase::run().

File: alica_tests/include/test_alica.h [hunk 1]
```diff
@@ -74,12 +74,13 @@ class AlicaTestFixture : public AlicaTestFixtureBase
...
         ac->setWorldModel<alicaTests::TestWorldModel>();
         ac->setTimerFactory<alicaRosTimer::AlicaRosTimerFactory>();
-        ae = AlicaTestsEngineGetter::getEngine(ac);
-        const_cast<IAlicaCommunication&>(ae->getCommunicator()).startCommunication();
         spinner->start();
         creators = {std::make_unique<alica::ConditionCreator>(), std::make_unique<alica::UtilityFunctionCreator>(),
                 std::make_unique<alica::ConstraintCreator>(), std::make_unique<alica::BehaviourCreator>(), std::make_unique<alica::PlanCreator>()};
-        EXPECT_TRUE(ae->init(std::move(creators)));
+
+        EXPECT_EQ(0,ac->init(std::move(creators)));
```
Comment (bjoernschroeder) [inline alica_tests/include/test_alica.h]: Initializing the context instead of the engine leads to starting the engine and failing some tests, which rely on an engine that is not running.

Comment (triccyx) [reply alica_tests/include/test_alica.h]: @bjoernschroeder 
Due to the change in AlicaContext::AlicaContext and AlicaContext::Init the flow is now:
- AlicaContext::AlicaContext no more create the AlicaEngine object
- AlicaContext::init Create and init the AlicaEngine object

so the
```c++
AlicaContext* ac = new alica::AlicaContext ...
ae=AlicaTestsEngineGetter::getEngine(ac);   Here the ae in nullptr
ae->init();
```
is no more possible

@bjoernschroeder if you agree I can change the two tests and keep only one endpoint

Comment (veerajsk) [reply alica_tests/include/test_alica.h]: Keeping only one endpoint will make the tests redundant I think. This test is odd, since it is looking at various internal engine APIs & making a bunch of assumptions about them & not following the normal initialization procedure. But I think the test can be modified to work on another plan instead of master. @bjoernschroeder can you check if it is possible to do this? Basically after engine is initialized in the normal way, switch to a state with a sub plan & do the asserts on that plan instead of master.

Comment (veerajsk) [reply alica_tests/include/test_alica.h]: @triccyx you can disable these tests until then

**Comment by triccyx** [comment]
@veerajsk 
Commented tests:
```       
#src/test/test_task_assignment.cpp 
#src/test/test_agent_dies.cpp
#src/test/test_config_change.cpp
#src/test/test_alica_scheduling.cpp  
#src/test/test_alica_tracing.cpp  
#src/test/test_failure_handling.cpp
```
 I have checked that  all these failed tests are related to the changing in AlicaContext::AlicaContext and AlicaContect::init
In all cases, the fixtures need to be changed.
The only test I have changed is the fixture in `alica_test.h`

**Comment by abhi1287** [comment]
@triccyx I am not in favour of disabling tests. Its a valid expectation from any test to delay starting engine to an appropriate time. For example, test might want to initialize some data before the behaviours or plans are run. 
Lets do the following:
1. in AlicaContext::init(), introduce a delayStart argument that defaults to false. If this flag is set, skip the engine->start() call.
2. in test_alica.h setup() call AlicaContext::init with delayStart set to true. Also remove all the duplicate code that will anyways happen in AlicaContext::init call.
3. Get the engine after init call and rest stays the same. you won't need to disable any test.

**Comment by triccyx** [comment]
@abhi1287 I have followed @veerajsk request as he will check the temporarily removed tests in the code review.

**Comment by triccyx** [comment]
Today I will implement @abhi1287 suggestion to try to reduce the number of failed tests. @veerajsk is it ok for you?

**Comment by triccyx** [comment]
Test problems after last fixes:
test_task_assignment --> too many entry points in master plan
test_agent_dies --> AlicaContext::setclock after AlicaContext::init()
test_config_change --> changes to config not allowed after initialization of AlicaContext
test_failure_handling --> Not analyzed
Supplementary test not analyzed

File: alica_engine/include/engine/AlicaContext.h [hunk 2]
```diff
@@ -501,6 +512,11 @@ bool AlicaContext::existSolver() const
 template <class TimerFactoryType, class... Args>
 void AlicaContext::setTimerFactory(Args&&... args)
 {
+    if (_initialized) {
```
Comment (veerajsk) [inline alica_engine/include/engine/AlicaContext.h:516]: need to include this check in all set methods. Missed out ones: `setTraceFactory`, `setWorldModel`

File: alica_engine/src/engine/AlicaContext.cpp [hunk 3]
```diff
@@ -16,14 +16,12 @@ constexpr uint32_t ALICA_CTX_GOOD = 0xaac0ffee;
 constexpr uint32_t ALICA_CTX_BAD = 0xdeaddead;
 constexpr int ALICA_LOOP_TIME_ESTIMATE = 33; // ms
 
-AlicaContext::AlicaContext(const AlicaContextParams& alicaContextParams)
+AlicaContext::AlicaContext(const AlicaContextParams alicaContextParams)
```
Comment (veerajsk) [inline alica_engine/src/engine/AlicaContext.cpp]: why is this changed from `const&` to value?

File: alica_engine/src/engine/AlicaContext.cpp [hunk 4]
```diff
@@ -32,27 +30,38 @@ AlicaContext::~AlicaContext()
...
-int AlicaContext::init(AlicaCreators&& creatorCtx)
+int AlicaContext::init(AlicaCreators&& creatorCtx, bool delayStart)
 {
     if (_initialized) {
         ALICA_WARNING_MSG("AC: Context already initialized.");
         return -1;
     }
 
+    if (!_communicator) {
+        AlicaEngine::abort("AC: Communicator not seted");
```
Comment (veerajsk) [inline alica_engine/src/engine/AlicaContext.cpp]: `seted` -> `set`

File: alica_engine/src/engine/AlicaContext.cpp [hunk 5]
```diff
@@ -32,27 +30,38 @@ AlicaContext::~AlicaContext()
...
+    }
+    if (!_timerFactory) {
+        AlicaEngine::abort("AC: TimerFactory not seted");
+    }
+
+    _engine = std::make_unique<AlicaEngine>(*this, _configRootNode, _alicaContextParams);
+
     if (_communicator) {
```
Comment (veerajsk) [inline alica_engine/src/engine/AlicaContext.cpp]: check is redundant because of the assert above

File: alica_engine/src/engine/AlicaContext.cpp [hunk 6]
```diff
@@ -32,27 +30,38 @@ AlicaContext::~AlicaContext()
...
         _communicator->startCommunication();
     }
 
     if (_engine->init(std::move(creatorCtx))) {
-        _engine->start();
         _initialized = true;
+        if (!delayStart)
+            _engine->start();
```
Comment (veerajsk) [inline alica_engine/src/engine/AlicaContext.cpp:60]: enclose even single line `if`s inside `{}` as a good coding guideline followed throughtout the rest of the engine

File: alica_engine/include/engine/AlicaContext.h [hunk 7]
```diff
@@ -202,7 +202,7 @@ class AlicaContext
      *
      * @see AlicaCreators
      */
-    int init(AlicaCreators&& creatorCtx);
+    int init(AlicaCreators&& creatorCtx, bool delayStart = false);
```
Comment (veerajsk) [inline alica_engine/include/engine/AlicaContext.h:206]: add a comment that `delayStart` is mainly for tests because there is no way to start the engine via the context otherwise.

File: alica_engine/include/engine/AlicaEngine.h [hunk 8]
```diff
@@ -49,7 +48,7 @@ class AlicaEngine
 
     // Parameter Access:
     bool getStepEngine() const;
-    bool maySendMessages() const { return _maySendMessages; }
+    const bool& maySendMessages() const { return _maySendMessages; }
```
Comment (veerajsk) [inline alica_engine/include/engine/AlicaEngine.h]: returning `const bool&` is not necessary, somewhat confusing & less performant. Would rather just return by value.

File: alica_engine/src/engine/AlicaEngine.cpp [hunk 9]
```diff
@@ -35,23 +35,22 @@ void AlicaEngine::abort(const std::string& msg)
...
         , _configChangeListener(config)
         , _stepCalled(false)
-        , _stepEngine(stepEngine)
+        , _stepEngine(alicaContextParams.stepEngine)
         , _log(this)
-        , _modelManager(_planRepository, _configChangeListener, configPath)
-        , _masterPlan(_modelManager.loadPlanTree(masterPlanName))
-        , _roleSet(_modelManager.loadRoleSet(roleSetName))
-        , _teamManager(this, agentID)
+        , _modelManager(_planRepository, _configChangeListener, alicaContextParams.configPath)
```
Comment (veerajsk) [inline alica_engine/src/engine/AlicaEngine.cpp]: the initialization order is determined by the order in which the members are declared, therefore it would improve readability a lot if the order here is same as the declaration order in the header file.

File: alica_engine/include/engine/AlicaEngine.h [hunk 10]
```diff
@@ -134,18 +133,19 @@ class AlicaEngine
     AlicaContext& _ctx;
     PlanRepository _planRepository;
     ModelManager _modelManager;
-    const Plan* _masterPlan; /**< Pointing to the top level plan of the loaded ALICA program.*/
-    const RoleSet* _roleSet; /**< Pointing to the current set of known roles.*/
+    const Plan* _masterPlan{nullptr}; /**< Pointing to the top level plan of the loaded ALICA program.*/
```
Comment (veerajsk) [inline alica_engine/include/engine/AlicaEngine.h]: would prefer to have all initialization in one place to avoid having to look back & forth while reading the code, so best to move it to the cpp file

File: alica_engine/include/engine/TeamObserver.h [hunk 11]
```diff
@@ -27,7 +30,8 @@ class SimplePlanTree;
 class TeamObserver
 {
 public:
-    TeamObserver(AlicaEngine* ae);
+    TeamObserver(Logger& logger, IRoleAssignment& roleAssigment, const bool& maySendMessages, const IAlicaCommunication& communicator, const AlicaClock& clock,
```
Comment (veerajsk) [inline alica_engine/include/engine/TeamObserver.h]: `const bool&` -> `bool`

File: alica_engine/include/engine/TeamObserver.h [hunk 12]
```diff
@@ -48,9 +52,14 @@ class TeamObserver
     void cleanOwnSuccessMarks(RunningPlan* root);
     std::unique_ptr<SimplePlanTree> sptFromMessage(AgentId agent, const IdGrp& ids, AlicaTime time) const;
 
-    AlicaEngine* _ae;
-    TeamManager& _tm;
     Agent* _me;
```
Comment (veerajsk) [inline alica_engine/include/engine/TeamObserver.h]: is dependent on team manager, so should be declared after it

File: alica_engine/src/engine/TeamObserver.cpp [hunk 13]
```diff
@@ -25,12 +25,18 @@ using std::map;
...
+        const AlicaClock& clock, const PlanRepository& planRepository, TeamManager& teamManager)
+        : _logger(logger)
+        , _roleAssignment(roleAssigment)
+        , _maySendMessages(maySendMessages)
+        , _communicator(communicator)
+        , _clock(clock)
+        , _planRepository(planRepository)
+        , _tm(teamManager)
 {
     _me = _tm.editLocalAgent();
```
Comment (veerajsk) [inline alica_engine/src/engine/TeamObserver.cpp]: once `_me` is declared after `_tm`, this can be moved to the initialization list

File: alica_tests/src/test/test_agent_dies.cpp [hunk 14]
```diff
@@ -61,7 +61,7 @@ TEST_F(AlicaEngineAgentDiesTest, AgentIsRemoved)
 {
     ASSERT_NO_SIGNAL
 
-    acs[0]->setClock<TestClock>();
+    acs[0]->setClock<TestClock>(); // LUCA setclock must be done befor ac.init
```
Comment (veerajsk) [inline alica_tests/src/test/test_agent_dies.cpp]: you could add a `virtual getClock(int agentNumber)` in `AlicaTestMultiAgentFixture` that returns the `AlicaClock()`, set the clock in `Setup()` & override it in `AlicaEngineAgentDiesTest` to return the `TestClock()`

File: alica_tests/src/test/test_alica_scheduling.cpp [hunk 15]
```diff
@@ -24,7 +24,7 @@ class AlicaSchedulingPlan : public AlicaSchedulingTestFixture
 
 TEST_F(AlicaSchedulingPlan, scheduling)
 {
-    ae->start();
+    ae->start(); // Missing in fixture  ac->init
```
Comment (veerajsk) [inline alica_tests/src/test/test_alica_scheduling.cpp]: I guess this is fixed now & working?

File: alica_tests/src/test/test_assignment.cpp [hunk 16]
```diff
@@ -34,6 +34,16 @@ TEST(Assignment, RobotsInserted)
...
 
+    ASSERT_TRUE(ac->isValid());
+    ac->setCommunicator<alicaDummyProxy::AlicaDummyCommunication>();
+    ac->setWorldModel<alicaTests::TestWorldModel>();
+    ac->setTimerFactory<alicaRosTimer::AlicaRosTimerFactory>();
+
+    alica::AlicaCreators creators = {std::make_unique<alica::ConditionCreator>(), std::make_unique<alica::UtilityFunctionCreator>(),
+            std::make_unique<alica::ConstraintCreator>(), std::make_unique<alica::BehaviourCreator>(), std::make_unique<alica::PlanCreator>()};
+
+    EXPECT_EQ(0, ac->init(std::move(creators)));
```
Comment (veerajsk) [inline alica_tests/src/test/test_assignment.cpp:45]: need to delay the start of the engine here

File: alica_tests/src/test/test_config_change.cpp [hunk 17]
```diff
@@ -18,7 +18,7 @@ TEST_F(AlicaNotInitialized, TestUpdatingComponents)
     alica::AlicaCreators creators(std::make_unique<alica::ConditionCreator>(), std::make_unique<alica::UtilityFunctionCreator>(),
             std::make_unique<alica::ConstraintCreator>(), std::make_unique<alica::BehaviourCreator>(), std::make_unique<alica::PlanCreator>());
 
-    EXPECT_TRUE(ac->setOption<bool>("Alica.SilentStart", false));
+    EXPECT_TRUE(ac->setOption<bool>("Alica.SilentStart", false)); // engine not yet created so AlicaContext::reload gives error
```
Comment (veerajsk) [inline alica_tests/src/test/test_config_change.cpp]: but the engine is constructed in `AlicaTestNotInitializedFixture`. Is this test still failing?

File: alica_engine/src/engine/AlicaContext.cpp (hunk 4 above)
Comment (triccyx) [reply alica_engine/src/engine/AlicaContext.cpp]: I need to review irregular verbs

File: alica_engine/include/engine/AlicaEngine.h (hunk 8 above)
Comment (triccyx) [reply alica_engine/include/engine/AlicaEngine.h]: leftover from some previous tests

File: alica_engine/include/engine/TeamObserver.h (hunk 11 above)
Comment (triccyx) [reply alica_engine/include/engine/TeamObserver.h]: I think this can be necessary for  `AlicaEngine::reload`
```c++
void AlicaEngine::reload(const YAML::Node& config)
{
    ...
    _maySendMessages = !config["Alica"]["SilentStart"].as<bool>();
   ...
}
```

File: alica_tests/src/test/test_assignment.cpp (hunk 16 above)
Comment (triccyx) [reply alica_tests/src/test/test_assignment.cpp:45]: I think it is all right here. The test is working. Note that `0` means all right

File: alica_tests/src/test/test_config_change.cpp (hunk 17 above)
Comment (triccyx) [reply alica_tests/src/test/test_config_change.cpp]: Yes still failing, the setOption should be done before AlicaContext::init so I changed the fixture;
also,
this piece of code needs to be removed for me (the _engine doesn't still exist here):
```c++
AlicaContext::setOptions()
{

 if (_initialized) {
        ALICA_WARNING_MSG("AC: Context already initialized. setOption not possibile.");
        return false;
    }
...
    //to be removed
    if (reload) {
        reloadConfig();
    }
...
}
```

Finally, some tests need to be split.

Comment (veerajsk) [reply alica_tests/src/test/test_config_change.cpp]: the changes you have now made look fine to me

Comment (triccyx) [reply alica_tests/src/test/test_config_change.cpp]: I have changed and now setOption can be executed also after AlicaContext::Init.

File: alica_tests/src/test/test_agent_dies.cpp (hunk 14 above)
Comment (triccyx) [reply alica_tests/src/test/test_agent_dies.cpp]: I set the clock in AlicaTestMultiAgentFixture::Setup() before init all the rest is already done in the test file

**Comment by triccyx** [comment]
@veerajsk @abhi1287 now all the tests work. If you can, could you review them? I have made some modifications, minor in my opinion.
Also, note that as @veerajsk suggested the behaviour of AlicaContext::setOption is slightly changed.
`I haven't still checked supplementary tests.`

File: alica_engine/include/engine/AlicaContext.h [hunk 18]
```diff
@@ -172,7 +172,7 @@ class AlicaContext
      * @note This is the main alica api class
      * @note Can throw an exception if some necessary configurations dont exist in the config file.
      */
-    AlicaContext(const AlicaContextParams& alicaContextParams);
+    AlicaContext(const AlicaContextParams alicaContextParams);
```
Comment (abhi1287) [inline alica_engine/include/engine/AlicaContext.h]: could still be taken by reference

File: alica_engine/include/engine/AlicaContext.h [hunk 19]
```diff
@@ -191,7 +191,7 @@ class AlicaContext
      *
      * @see AlicaCreators
      */
-    [[deprecated("call init(std::move(creators)) instead")]] int init(AlicaCreators& creatorCtx);
+    [[deprecated("call init(std::move(creators)) instead")]] int init(AlicaCreators& creatorCtx, bool delayStart = false);
```
Comment (abhi1287) [inline alica_engine/include/engine/AlicaContext.h]: no need to make changes to deprecated api.

File: alica_engine/include/engine/AlicaContext.h (hunk 7 above)
Comment (abhi1287) [reply alica_engine/include/engine/AlicaContext.h:206]: i don't see the comment.

File: alica_engine/include/engine/AlicaEngine.h [hunk 20]
```diff
@@ -49,7 +48,7 @@ class AlicaEngine
 
     // Parameter Access:
     bool getStepEngine() const;
-    bool maySendMessages() const { return _maySendMessages; }
+    const bool maySendMessages() const { return _maySendMessages; }
```
Comment (abhi1287) [inline alica_engine/include/engine/AlicaEngine.h]: no point of const here on return by value

**Comment by triccyx** [comment]
Fixed supplementary tests

File: alica_engine/include/engine/TeamObserver.h (hunk 11 above)
Comment (abhi1287) [inline alica_engine/include/engine/TeamObserver.h]: can rather pass the config here rather than const bool&

File: alica_engine/src/engine/AlicaContext.cpp [hunk 21]
```diff
@@ -32,27 +30,37 @@ AlicaContext::~AlicaContext()
...
+int AlicaContext::init(AlicaCreators& creatorCtx, bool delayStart)
 {
     AlicaCreators creators(std::move(creatorCtx.conditionCreator), std::move(creatorCtx.utilityCreator), std::move(creatorCtx.constraintCreator),
             std::move(creatorCtx.behaviourCreator), std::move(creatorCtx.planCreator));
-    return init(std::move(creators));
+    return init(std::move(creators), delayStart);
 }
 
-int AlicaContext::init(AlicaCreators&& creatorCtx)
+int AlicaContext::init(AlicaCreators&& creatorCtx, bool delayStart) // delayStarted==true is only used for testing purpose
```
Comment (abhi1287) [inline alica_engine/src/engine/AlicaContext.cpp]: comment should be in header, in doxygen compatible format

File: alica_engine/src/engine/AlicaContext.cpp [hunk 22]
```diff
@@ -32,27 +30,37 @@ AlicaContext::~AlicaContext()
...
+        AlicaEngine::abort("AC: TimerFactory not set");
     }
 
+    _engine = std::make_unique<AlicaEngine>(*this, _configRootNode, _alicaContextParams);
+
+    _communicator->startCommunication();
+
     if (_engine->init(std::move(creatorCtx))) {
-        _engine->start();
         _initialized = true;
```
Comment (abhi1287) [inline alica_engine/src/engine/AlicaContext.cpp]: set this flag after the next if condition.

File: alica_engine/src/engine/AlicaEngine.cpp [hunk 23]
```diff
@@ -100,6 +104,8 @@ bool AlicaEngine::init(AlicaCreators&& creatorCtx)
     _syncModul.init();
     _variableSyncModule->init();
     _auth.init();
+
+    _initialized = true;
```
Comment (abhi1287) [inline alica_engine/src/engine/AlicaEngine.cpp:109]: unset this flag in terminate.

File: alica_tests/include/test_alica.h [hunk 24]
```diff
@@ -53,6 +53,7 @@ class AlicaTestFixture : public AlicaTestFixtureBase
     alica::AlicaCreators creators;
 
 protected:
+    bool delayStart{false};
```
Comment (abhi1287) [inline alica_tests/include/test_alica.h]: this should be a virtual function that tests can override. Just like other following functions.

File: alica_tests/include/test_alica.h [hunk 25]
```diff
@@ -74,12 +75,13 @@ class AlicaTestFixture : public AlicaTestFixtureBase
...
-        ae = AlicaTestsEngineGetter::getEngine(ac);
-        const_cast<IAlicaCommunication&>(ae->getCommunicator()).startCommunication();
         spinner->start();
         creators = {std::make_unique<alica::ConditionCreator>(), std::make_unique<alica::UtilityFunctionCreator>(),
                 std::make_unique<alica::ConstraintCreator>(), std::make_unique<alica::BehaviourCreator>(), std::make_unique<alica::PlanCreator>()};
-        EXPECT_TRUE(ae->init(std::move(creators)));
+
+        EXPECT_EQ(0, ac->init(std::move(creators), delayStart));
+        ae = AlicaTestsEngineGetter::getEngine(ac);
+        const_cast<IAlicaCommunication&>(ae->getCommunicator()).startCommunication();
```
Comment (abhi1287) [inline alica_tests/include/test_alica.h]: ac->init starts the communicator so you don't need to do it again

File: alica_tests/include/test_alica.h [hunk 26]
```diff
@@ -145,10 +163,12 @@ class AlicaTestMultiAgentFixture : public AlicaTestMultiAgentFixtureBase
             ac->setCommunicator<alicaDummyProxy::AlicaDummyCommunication>();
             ac->setWorldModel<alicaTests::TestWorldModel>();
             ac->setTimerFactory<alicaRosTimer::AlicaRosTimerFactory>(*cbQueues.back());
+            if (useTestClock)
```
Comment (abhi1287) [inline alica_tests/include/test_alica.h]: use {}

File: alica_tests/include/test_alica.h [hunk 27]
```diff
@@ -112,12 +114,28 @@ class AlicaTestMultiAgentFixtureBase : public ::testing::Test
...
+};
+
 class AlicaTestMultiAgentFixture : public AlicaTestMultiAgentFixtureBase
 {
 private:
     alica::AlicaCreators creators;
 
 protected:
+    bool delayStart{true};
+    bool useTestClock{false};
```
Comment (abhi1287) [inline alica_tests/include/test_alica.h]: virtual function

File: alica_engine/include/engine/AlicaContext.h (hunk 19 above)
Comment (triccyx) [reply alica_engine/include/engine/AlicaContext.h]: ok

File: alica_engine/include/engine/AlicaEngine.h (hunk 20 above)
Comment (triccyx) [reply alica_engine/include/engine/AlicaEngine.h]: leftover

File: alica_engine/include/engine/AlicaContext.h (hunk 18 above)
Comment (triccyx) [reply alica_engine/include/engine/AlicaContext.h]: @abhi1287 
We have a rvalue problem here:
TestContext.cpp:17
```c++
: AlicaContext(AlicaContextParams(agentName, configPath, roleSetName, masterPlanName, stepEngine, agentID))
```
I would avoid using:
```c++
 : AlicaContext(std::forward<AlicaContextParams>(AlicaContextParams(agentName, configPath, roleSetName, masterPlanName, stepEngine, agentID)))
        , _initCalled(false)
```
or we can 
```c++
AlicaContext(const AlicaContextParams&& ...
```
But we change the interface.
However, as the copy is done once I would keep it as it is now.

File: alica_engine/include/engine/TeamObserver.h (hunk 11 above)
Comment (triccyx) [reply alica_engine/include/engine/TeamObserver.h]: @abhi1287  With this modification, we have 5 failures in tests. If we want to do this I need to debug the tests.

File: alica_engine/include/engine/AlicaContext.h (hunk 18 above)
Comment (triccyx) [reply alica_engine/include/engine/AlicaContext.h]: In the end, I decided to add a new  move constructor :
```c++
AlicaContext(const AlicaContextParams&& alicaContextParams);
````

File: alica_engine/include/engine/TeamObserver.h (hunk 11 above)
Comment (triccyx) [reply alica_engine/include/engine/TeamObserver.h]: ok find the problem

File: alica_engine/include/engine/AlicaContext.h [hunk 28]
```diff
@@ -173,6 +173,7 @@ class AlicaContext
      * @note Can throw an exception if some necessary configurations dont exist in the config file.
      */
     AlicaContext(const AlicaContextParams& alicaContextParams);
+    AlicaContext(const AlicaContextParams&& alicaContextParams);
```
Comment (veerajsk) [inline alica_engine/include/engine/AlicaContext.h]: const && does not make sense in general & should never be used. It essentially is a move reference from which you can't move because its const. Please change this back to const &.

Comment (veerajsk) [reply alica_engine/include/engine/AlicaContext.h]: there should not be any problem in TestContext either with this. I don't understand what the problem is with const &

File: alica_engine/src/engine/TeamObserver.cpp [hunk 29]
```diff
@@ -124,7 +122,8 @@ void TeamObserver::close()
  */
 void TeamObserver::doBroadCast(const IdGrp& msg) const
 {
-    if (!_maySendMessages) {
+    bool maySendMessages = !_config["Alica"]["SilentStart"].as<bool>();
```
Comment (veerajsk) [inline alica_engine/src/engine/TeamObserver.cpp]: - need a reload function for the config
- can load this value whenever the config changes instead of doing it here

File: alica_tests/include/test_alica.h [hunk 30]
```diff
@@ -145,10 +162,13 @@ class AlicaTestMultiAgentFixture : public AlicaTestMultiAgentFixtureBase
             ac->setCommunicator<alicaDummyProxy::AlicaDummyCommunication>();
             ac->setWorldModel<alicaTests::TestWorldModel>();
             ac->setTimerFactory<alicaRosTimer::AlicaRosTimerFactory>(*cbQueues.back());
+            if (getUseTestClock()) {
+                ac->setClock<TestClock>();
+            }
+            ac->init(std::move(creators), getDelayStart());
             alica::AlicaEngine* ae = AlicaTestsEngineGetter::getEngine(ac);
             const_cast<IAlicaCommunication&>(ae->getCommunicator()).startCommunication();
```
Comment (veerajsk) [inline alica_tests/include/test_alica.h]: can this be removed as well? ac->init starts the communication anyway

File: alica_tests/include/test_alica.h [hunk 31]
```diff
@@ -53,6 +53,7 @@ class AlicaTestFixture : public AlicaTestFixtureBase
     alica::AlicaCreators creators;
 
 protected:
+    virtual bool getDelayStart() { return false; }
```
Comment (veerajsk) [inline alica_tests/include/test_alica.h]: shouldn't this return true since ae->start() was not being called earlier

File: alica_tests/src/test/test_assignment.cpp (hunk 16 above)
Comment (veerajsk) [reply alica_tests/src/test/test_assignment.cpp:45]: my point was that the test earlier did not start the alica engine

**Review by veerajsk — COMMENTED** [review]
Why can't the engine start be delayed in every test fixture which would make the code equivalent to what it was before? That way none of the tests need to be updated. Right now it is difficult to review because some tests start the engine while others don't & therefore have to go through all of them to know if they will still work.

Imo, the only changes required would be to call ac->init() with delayed start before AlicaTestsEngineGetter::getEngine & setting the clock for AlicaEngineAgentDiesTest.

File: alica_engine/include/engine/AlicaContext.h (hunk 28 above)
Comment (triccyx) [reply alica_engine/include/engine/AlicaContext.h]: I don't know or even noticed that "c++ rule" about bound const& to lvalue without using the move constructor.
However, I have done this modification because the test won't link. I will check again this evening.  
These work
`AlicaContext(const AlicaContextParams alicaContextParams);` 
`AlicaContext(const AlicaContextParams&& alicaContextParams);` 
Perhaps I have just done an error aligning the method in .cpp.

Comment (triccyx) [reply alica_engine/include/engine/AlicaContext.h]: Ok bug found something left in the old build

File: alica_engine/src/engine/TeamObserver.cpp (hunk 29 above)
Comment (triccyx) [reply alica_engine/src/engine/TeamObserver.cpp]: ok

File: alica_tests/include/test_alica.h (hunk 31 above)
Comment (triccyx) [reply alica_tests/include/test_alica.h]: In this way, the behaviour is "old behaviour" for all tests except only the ones that need a delay.

**Review by veerajsk — COMMENTED** [review]
Looks fine except for my earlier comment of delaying the engine start for all tests

**Comment by triccyx** [comment]
@veerajsk I have done some work on delaystart. Now all but 3 tests need delaystart=false.
Looking at the tests kind it looks good. But I haven't all the knowledge to judge.

**Review by abhi1287 — APPROVED** [review]
